dear_petition/petition/tests/test_models.py

```diff
@@ -148,6 +213,106 @@ def test_refresh_record_from_data():
+
+    assert len(stored_offense_records) == len(raw_data_offense_records)
+    assert len(stored_actions) == len(raw_data_actions)
+    for action in stored_actions:
+        assert action in raw_data_actions
+
+    # Offense Record's Record key has changed on the raw data to only having 1 record.
+    # lets make sure there is only one record being stored and this is it.
+    new_o_record_code = "9999"
+    updated_o_records = {
+        "Records": [
+            {
+                "Action": "CONVICTED",
+                "Description": "IMPROPER EQUIP - SPEEDOMETER",
+                "Severity": "INFRACTION",
+                "Law": "G.S. 20-123.2",
+                "Code": new_o_record_code,
+            },
+        ]
+    }
+    ciprs_record.data.get("Offense Record").update(updated_o_records)
+    ciprs_record.refresh_record_from_data()
+
+    raw_data_offense_records = ciprs_record.data.get("Offense Record").get("Records")
+    stored_offense_records = list(ciprs_record.offenses.first().offense_records.all())
+    assert len(stored_offense_records) == len(raw_data_offense_records)
+    assert int(new_o_record_code) == stored_offense_records[0].code
+
+    # Now lets add three offenses and call refresh
+    o_record_code_list = ["1111", "2222", "3333", "4444"]
+    o_record_int_code_list = [int(code) for code in o_record_code_list]
+    updated_o_records = {
+        "Records": [
+            {
+                "Action": "CONVICTED",
+                "Description": "IMPROPER EQUIP - SPEEDOMETER",
+                "Severity": "INFRACTION",
+                "Law": "G.S. 20-123.2",
+                "Code": o_record_code_list[0],
+            },
+            {
+                "Action": "CHARGED",
+                "Description": "SPEEDING(80 mph in a 65 mph zone)",
+                "Severity": "INFRACTION",
+                "Law": "G.S. 20-141(B)",
+                "Code": o_record_code_list[1],
+            },
+            {
+                "Action": "ARRAIGNED",
+                "Description": "SPEEDING(80 mph in a 65 mph zone)",
+                "Severity": "INFRACTION",
+                "Law": "G.S. 20-141(B)",
+                "Code": o_record_code_list[2],
+            },
+            {
+                "Action": "CONVICTED",
+                "Description": "IMPROPER EQUIP - SPEEDOMETER",
+                "Severity": "INFRACTION",
+                "Law": "G.S. 20-123.2",
+                "Code": o_record_code_list[3],
+            },
+        ]
+    }
+    ciprs_record.data.get("Offense Record").update(updated_o_records)
+    ciprs_record.refresh_record_from_data()
+
+    raw_data_offense_records = ciprs_record.data.get("Offense Record").get("Records")
+    stored_codes = []
+    stored_offense_records = list(ciprs_record.offenses.first().offense_records.all())
+    for offense_record in stored_offense_records:
+        stored_codes.append(offense_record.code)
+    assert len(stored_offense_records) == len(raw_data_offense_records)
```

Comment:
Could you add  an `assert len(stored_codes) == len(o_record_int_code_list)` here, just to make sure that `stored_codes` and `o_record_int_code_list` really have the same elements?

Reply:
sure

Reply:
done